bitsail-connectors/connector-redis/src/main/java/com/bytedance/bitsail/connector/redis/core/AbstractPipelineProcessor.java

```diff
@@ -0,0 +1,419 @@
+            jedis.getClient(),
+            command.print(), ((Throwable) response).getMessage(), (Throwable) response);
+        
+        failureHandler.handle(command, (Throwable) response, this);
+      } else if (ttlResponse instanceof Throwable) {
+        LOG.warn("Set ttl error! " + ((Throwable) ttlResponse).getMessage(), (Throwable) ttlResponse);
+        failureHandler.handle(command, (Throwable) ttlResponse, this);
+      } else {
+        // success
+        this.successfulRecords.add(command);
+      }
+    }
+  }
+
+  @Override
+  public void postExecute(List<Command> curRequests, Throwable failure) {
+    for (Command command : curRequests) {
+      handleUnexpectedFailedRecord(command, failure);
+    }
+  }
+
+  @Override
+  public void postExecuteWithRetry(List<Command> curRequests, Throwable failure) {
+    for (Command command : curRequests) {
+      handleNeedRetriedRecords(command, failure);
+    }
+  }
+
+  @Override
+  public boolean needRetry() {
+
+    if (CollectionUtils.isNotEmpty(unexpectedFailedRecords)) {
+      throw new UnexpectedException(printErrorMessage(), unexpectedFailedRecords.get(0).getSecond());
+    }
+    boolean needRetry = attemptNumber.get() < RedisWriter.MAX_ATTEMPT_NUM
+        && CollectionUtils.isNotEmpty(needRetriedRecords);
+
+    if (!needRetry) {
+      if (needRetriedRecords.isEmpty()) {
+        if (hitLogSampling()) {
+          LOG.info("finished pipeline [{}] records, attempt number:[{}], processor id:[{}]",
+              commandSize, attemptNumber.get(), processorId);
+        }
+      } else {
+        Throwable throwable = needRetriedRecords.get(0).getSecond();
+        if (throwable instanceof JedisDataException) {
+          LOG.error("pipeline finally failed after {} attempts. {}", attemptNumber.get(), printErrorMessage());
+          throw new BitSailException(JedisPluginErrorCode.PIPELINE_ERROR, printErrorMessage());
+        } else {
+          throw new UnexpectedException(printErrorMessage(), unexpectedFailedRecords.get(0).getSecond());
+        }
+      }
+    } else {
+      LOG.error("pipeline temporarily failed after {} attempts. {}", attemptNumber.get(), printErrorMessage());
+    }
+    return needRetry;
+  }
+
+  @Override
+  public void addInitialCommand(Command command) {
+    this.requests.add(command);
+  }
+
+  @Override
+  public void handleUnexpectedFailedRecord(Command command, Throwable throwable) {
+    this.unexpectedFailedRecords.add(Pair.newPair(command, throwable));
+  }
+
+  @Override
+  public void handleDirtyRecords(Command command, JedisDataException exception) {
+    this.dirtyRecords.add(Pair.newPair(command, exception));
+  }
+
+  @Override
+  public void handleNeedRetriedRecords(Command command, Throwable exception) {
+    this.needRetriedRecords.add(Pair.newPair(command, exception));
+  }
+
+  /**
+   * Submit command to pipeline.
+   */
+  protected void submitCommand(Command command) {
+    switch (command.getJedisCommand()) {
```

Comment:
not all operations matches what we mention in the doc. Do we support all of these operations?

Reply:
this is all operations in the doc 
![image](https://user-images.githubusercontent.com/47937942/200472469-46cdd7fd-42c0-4a4d-8a70-93ef21fe415e.png)
Sorted set is  HSET    
Sorted set is ZADD
Set is SADD
String is SET